src/pretty/to_fixed/mod.rs

```diff
@@ -0,0 +1,450 @@
+    // The first value this approximation fails for is 2^1651 which is just greater than 10^297.
+    assert!(e >= 0);
+    assert!(e <= 1650);
+    ((e as u32) * 78913) >> 18
+}
+
+fn index_for_exponent(e: u32) -> u32 {
+    (e + 15) / 16
+}
+
+const POW10_ADDITIONAL_BITS: u32 = 120;
+
+fn pow10_bits_for_index(idx: u32) -> u32 {
+    16 * idx + POW10_ADDITIONAL_BITS
+}
+
+fn length_for_index(idx: u32) -> u32 {
+    // +1 for ceil, +16 for mantissa, +8 to round up when dividing by 9
+    (log10_pow2(16 * idx as i32) + 1 + 16 + 8) / 9
+}
+
+fn umul256(a: u128, b_hi: u64, b_lo: u64) -> (u128, u128) {
+    let a_lo = a as u64;
+    let a_hi = (a >> 64) as u64;
+
+    let b00 = (a_lo as u128) * (b_lo as u128);
+    let b01 = (a_lo as u128) * (b_hi as u128);
+    let b10 = (a_hi as u128) * (b_lo as u128);
+    let b11 = (a_hi as u128) * (b_hi as u128);
+
+    let b00_lo = b00 as u64;
+    let b00_hi = (b00 >> 64) as u64;
+
+    let mid1 = b10 + b00_hi as u128;
+    let mid1_lo = (mid1) as u64;
+    let mid1_hi = (mid1 >> 64) as u64;
+
+    let mid2 = b01 + mid1_lo as u128;
+    let mid2_lo = (mid2) as u64;
+    let mid2_hi = (mid2 >> 64) as u64;
+
+    let p_hi = b11 + mid1_hi as u128 + mid2_hi as u128;
+    let p_lo = ((mid2_lo as u128) << 64) | b00_lo as u128;
+
+    (p_hi, p_lo)
+}
+
+// Returns the high 128 bits of the 256-bit product of a and b.
+fn umul256_hi(a: u128, b_hi: u64, b_lo: u64) -> u128 {
+    // Reuse the umul256 implementation.
+    // Optimizers will likely eliminate the instructions used to compute the
+    // low part of the product.
+    let (hi, _lo) = umul256(a, b_hi, b_lo);
+    hi
+}
+
+// Unfortunately, gcc/clang do not automatically turn a 128-bit integer division
+// into a multiplication, so we have to do it manually.
+fn uint128_mod1e9(v: u128) -> u32 {
+    // After multiplying, we're going to shift right by 29, then truncate to uint32_t.
+    // This means that we need only 29 + 32 = 61 bits, so we can truncate to uint64_t before shifting.
+    let multiplied = umul256_hi(v, 0x89705F4136B4A597, 0x31680A88F8953031) as u64;
+
+    // For uint32_t truncation, see the mod1e9() comment in d2s_intrinsics.rs
+    let shifted = (multiplied >> 29) as u32;
+
+    (v as u32).wrapping_sub(1000000000u32.wrapping_mul(shifted))
+}
+
+// Best case: use 128-bit type.
+fn mul_shift_mod1e9(m: u64, mul: &[u64; 3], j: i32) -> u32 {
+    let b0 = m as u128 * mul[0] as u128; // 0
+    let b1 = m as u128 * mul[1] as u128; // 64
+    let b2 = m as u128 * mul[2] as u128; // 128
+
+    assert!(j >= 128);
+    assert!(j <= 180);
+    // j: [128, 256)
+    let mid = b1 + ((b0 >> 64) as u64) as u128; // 64
+    let s1 = b2 + ((mid >> 64) as u64) as u128; // 128
+    uint128_mod1e9(s1 >> (j - 128))
+}
+
+// Convert `digits` to decimal and write the last 9 decimal digits to result.
+// If `digits` contains additional digits, then those are silently ignored.
+unsafe fn append_nine_digits(mut digits: u32, result: *mut u8) {
+    // #ifdef RYU_DEBUG
```

Comment:
Should we keep this?

Reply:
This is left over from the port, will remove it :)